---
title: "FacileDataSet Overview"
author: "Steve Lianoglou"
date: "`r Sys.Date()`"
package: "`r BiocStyle::pkg_ver('FacileDataSet')`"
output:
  html_document:
    code_folding: show
    fig_caption: yes
    self_contained: yes
    highlight: pygments
    fig_width: 5
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
vignette: >
  %\VignetteIndexEntry{FacileDataSet Overview}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding[utf8]{inputenc}
---

```{r setup, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  echo=TRUE, warning=FALSE, message=FALSE, error=FALSE) #, dpi=150)
library(ggplot2)
library(cowplot)
```

# Overview

Briefly, there are two pieces of software defined in this package.

1. A "FacileData Access API" that provides a fluent interface that enables
   analysts to more naturally query and retrieve data for general exploratory
   data analyses; and
2. A high-throughput genomics assay container called a `FacileDataSet`. The
   `FacileDataSet` stores  multi-assay and multi-dataset genomics data on disk
   (out of memory) and provides efficient retrieval of these data through using
   the *FacileData API*. A single `FacileDataSet` can be used to house *all* of
   the RNA-seq, microarray, RPPA, etc. data from the TCGA, for example.

# Background

The [Bioconductor ecosystem][bioc] provides a number of data containers, like
the [`SummarizedExperiment`][SummarizedExperiment] to store high-throughput
genomics data and the metadata that goes along with it. These containers provide
a consistent way for users to access the different aspects of the data that are
required to successfully analyze genomics data and have served the community
well for a long time, *especially* in the context of single-experiment projects.

Storing data in these containers, however, is sub-optimal in a few situations:

1. When the analyst wants to perform more general exploratory data analysis
   techniques over the data (like simple scatterplots, boxplots, etc.); and
2. When the analyst is analyzing similar data over very large datasets, such as
   *the entirety* of the data from [The Cancer Genome Atlas][tcga] (TCGA),
   particularly due to the fact that these data containers want to store all of
   the data in RAM.
   
The fluent and efficient analysis of data from the TCGA poses even more problems
since a multitude of assays have been run over each sample, such as RNA-seq,
microarrays, exome-seq, etc. With all of this data in hand, analysts may want to
look at how gene expression compares to copy number, or miRNA expression, or ...

In these scenarios, a single `SummarizedExperiment` is no longer up to the task
of storing these data since the "feature-space" (rows of the
`SummarizedExperiment`) must be equal, and the analyst is now forced to
coordinate access and merge data across a number of such containers: one for
each assay (generally speaking).

... continue ...

Data generation efforts, like the TCGA, are becoming more common. Other large consortia (ENCODE, ICGC, etc.) are generating and  in the public as well as private data that is now more common to generate
from large clinical trials underway in industry.


Bioinformaticians often find themselves trying to integrate heterogeneous
sources of data across many datasets as they analyze high throughput genomics
data in different ways, ie. exploratory data analysis, predictive modeling,
differential expression analysis, etc. The current tools that are made available
to help corral these data

Although bioconductor provides a number
of data assay containers, like a `SummarizedExperiment` and `ExpressionSet` to
help corale these types of data


[bioc]: http://bioconductor.org/
[SummarizedExperiment]: http://bioconductor.org/packages/SummarizedExperiment/
[tcga]: https://cancergenome.nih.gov/

It is often the
The `FacileDataSet` package defines this a container that stores multi-assy high throughput
genomics data across large datasets.

We will use the multi-assay, multi-dataset (indication) TCGA resource as a
data source to showcase how programming against a FacileDataSet faciliates
easier exploratory data analyses.

The `FacileTCGADataSet` provides the full RNA-seq expression, miRNA expresion,
and copy number data across all 11,000+ samples from the TCGA. These are the
number of normal/tumor samples available across all indications:

```{r tcga-sample-distribution}
library(FacileDataSet)
library(FacileTCGADataSet)
tcga <- FacileTCGADataSet()
distro <- samples(tcga) %>%
  with_sample_covariates(c("indication", "sample_type"))
ggplot(distro, aes(indication, fill=sample_type)) +
  geom_bar() +
  theme(axis.text.x=element_text(angle=90, hjust=1))
```



# FacileData Access API

The base interface is the `FacileDataStore`. There are no objects of this class,
it rather serves as an interface that should be implemented by a subclass, ie.
a `FacileDataSet`, or a `FacileSummarizedExperiment`. An object that
`is(x, "FacileDataStore") == TRUE` is assumed to implement the entire
*FacileData API*.

(unimplemented)
Broadly speaking, there are two classes of functions that are used to pull data
out of a `FacileDataSet`, the `fetch_*` and `with_*` functions. To a first
approximation, we can consider the `fetch_*` functions to be the lowest level
functions, and the `with_*` functions are higher-level accessors.

The `with_*` functions are *a bit* fast-and-loose. These are written to more
naturally combine in pipe (`%>%`) data manipulation chains and provide analysts
with a more fluent and succinct API with which to perform exploratory data
analyses.

For example, to retrieve normalized RNA-seq expression values for CXCL9 and GZMA
along with survival and sex information for the samples in the samples in the
TCGA bladder ("BLCA") indication, you can write:

```{r quick-blca-example}
features <- filter_features(tcga, name %in% c("CXCL9", "GZMA"))

blca.dat <- tcga %>%
  filter_samples(indication == "BLCA") %>% 
  with_assay_data(features, assay_name="rnaseq") %>% 
  with_sample_covariates(c("OS", "sex"))

ggplot(blca.dat, aes(CXCL9, GZMA, color=sex)) +
  geom_point()
```



# Exemplar Analyses

Provide a number of common anlaysis examples using the FacileTCGADataSet.

## Pan-cancer transcriptomics with RPPA

Should we try to reconstruct an analysis from Avi Ma'ayan's recent paper?:
http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005911


# Data Model

The pData data.frame is stored in the sample_covariate table in the internal
database using an [entity-attribute-value model][EAV].

[EAV]: https://en.wikipedia.org/wiki/Entity%E2%80%93attribute%E2%80%93value_model

