---
title: "FacileData API Overview"
author: "Steve Lianoglou"
date: "`r Sys.Date()`"
package: "`r BiocStyle::pkg_ver('FacileData')`"
output:
  html_document:
    code_folding: show
    fig_caption: yes
    self_contained: yes
    highlight: pygments
    fig_width: 5
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
    css: mystyles.css
  rmarkdown::html_vignette:
    css: mystyles.css
vignette: >
  %\VignetteIndexEntry{FacileData API Overview}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding[utf8]{inputenc}
---

```{r setup, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  echo=TRUE, warning=FALSE, message=FALSE, error=FALSE) #, dpi=150)
library(ggplot2)
# library(cowplot)
theme_set(theme_bw())
```

# Overview

The `FacileData` package was written to facilitate easier analysis of large,
multi-assay high-throughput genomics datasets. To this end, the `FacileData`
package provides two things:

1. A *FacileData Access API* that defines a fluent interface over multi-assay
   genomics datasets that fits into the [tidyverse][tidyverse]. This enables
   analysts to more naturally query and retrieve data for general exploratory
   data analysis; and
2. A reference implementation of a datastore that implements the
   *FacileData Access API* called a *FacileDataSet*. The `FacileDataSet`
   provides efficient storage and retrieval of arbitrarily large high-throughput
   genomics datasets. For example, a single `FacileDataSet` can be used to store
   *all* of the RNA-seq, microarray, RPPA, etc. data from the 
   [The Cancer Genome Atlas][tcga]. This singular `FacileDataSet` allows
   analysts easy access to arbitrary subsets of these data without having to
   load all of it into memory.

[tcga]: https://cancergenome.nih.gov/
[tidyverse]: https://www.tidyverse.org/

# Background

The [Bioconductor ecosystem][bioc] provides a number of data containers, like
the [`SummarizedExperiment`][SummarizedExperiment] to store high-throughput
genomics data and the metadata that goes along with it. These containers provide
a consistent way for users to access the different aspects of the data that are
required to successfully analyze genomics data and have served the community
well for a long time, *especially* in the context of single-experiment projects.

Storing data in these containers, however, is sub-optimal in a few situations:

1. When the analyst wants to perform more general exploratory data analysis
   techniques over the data (like simple scatterplots, boxplots, etc.). When
   the analyst is contrained by the data access API defined by a
   SummarizedExperiment, creating simple scatter and boxplots (for exmaple)
   can become quite tedious; and
2. When the analyst is analyzing similar data over very large datasets, such as
   *the entirety* of the data from [The Cancer Genome Atlas][tcga] (TCGA),
   particularly due to the fact that these data containers want to store all of
   the data in RAM.
   
The fluent and efficient analysis of data from the TCGA poses even more problems
since a multitude of assays have been run over each sample, such as RNA-seq,
microarrays, exome-seq, etc. With all of this data in hand, analysts may want to
look at how gene expression compares to copy number, or miRNA expression, etc.

In these scenarios, a single `SummarizedExperiment` is no longer up to the task
of storing these data since the "feature-space" (rows of the
`SummarizedExperiment`) of all the assays within a singular
`SummarizedExperiment` must be equal.

These problems manifest themselves when we consider even the simplest of
scenarios: consider processing an RNA-seq assay into two different feature
spaces, (1) gene-level counts; and (2) transcript-level counts. One would have
to create two separate `SummarizedExperiments` with the same columns and
`colData`, but different assay matrices given the different feature-spaces
(genes vs transcripts).

Using a single `FacileDataSet`, all of these data can be stored in the same
data container, and the *FacileData API* allows analysts to retrieve data from
each effortlessly.

[bioc]: http://bioconductor.org/
[SummarizedExperiment]: http://bioconductor.org/packages/SummarizedExperiment/

# FacileData API

The *FacileData API* defines a set of query and retrieval functions over
multi-assay genomics datasets that fits into the [tidyverse][tidyverse].

<div class="note">
To provide illustrative examples of the utility of the *FacileData API*,
we rely on the `FacileTCGADataSet` package, which provides the full RNA-seq
expression, miRNA expresion, and copy number data across all 33 indications 
(11,000+ samples) from the TCGA. Please see the `FacileTCGADataSet` package
for more information regarding how these data were put together, and how to
extend it.
</div>

As a teaser, we provide code snippets that show how to plot HER2 copy number
vs expression across the TCGA "BLCA" and "BRCA" indications using the=
`FacileDataSet`. We'll then compare that to how the same code might be written
using more traditional bioconductor containers.

```{r her2-cnv-expression}
library(FacileData)
library(FacileTCGADataSet)
tcga <- FacileTCGADataSet()

samples <- filter_samples(tcga, indication %in% c("BLCA", "BRCA"))
features <- filter_features(tcga, name == "ERBB2")

fdat <- samples %>% 
  with_assay_data(features, assay_name = "rnaseq", normalized = TRUE) %>%
  with_assay_data(features, assay_name = "cnv_score") %>% 
  with_sample_covariates(c("indication", "sex"))

ggplot(fdat, aes(cnv_score_ERBB2, ERBB2, color=sex)) +
  geom_point() +
  facet_wrap(~ indication)
```

**TODO: Write a fair traditional bioc-style implementation**

(use biobroom?)

```{r tcga-sample-distribution}
library(FacileDataSet)
library(FacileTCGADataSet)
tcga <- FacileTCGADataSet()
distro <- samples(tcga) %>%
  with_sample_covariates(c("indication", "sample_type"))
ggplot(distro, aes(indication, fill=sample_type)) +
  geom_bar() +
  theme(axis.text.x=element_text(angle=90, hjust=1))
```



# FacileData Access API

The base interface is the `FacileDataStore`. There are no objects of this class,
it rather serves as an interface that should be implemented by a subclass, ie.
a `FacileDataSet`, or a `FacileSummarizedExperiment`. An object that
`is(x, "FacileDataStore") == TRUE` is assumed to implement the entire
*FacileData API*.

(unimplemented)
Broadly speaking, there are two classes of functions that are used to pull data
out of a `FacileDataSet`, the `fetch_*` and `with_*` functions. To a first
approximation, we can consider the `fetch_*` functions to be the lowest level
functions, and the `with_*` functions are higher-level accessors.

The `with_*` functions are *a bit* fast-and-loose. These are written to more
naturally combine in pipe (`%>%`) data manipulation chains and provide analysts
with a more fluent and succinct API with which to perform exploratory data
analyses.

For example, to retrieve normalized RNA-seq expression values for CXCL9 and GZMA
along with survival and sex information for the samples in the samples in the
TCGA bladder ("BLCA") indication, you can write:

```{r quick-blca-example}
features <- filter_features(tcga, name %in% c("CXCL9", "GZMA"))

blca.dat <- tcga %>%
  filter_samples(indication == "BLCA") %>% 
  with_assay_data(features, assay_name="rnaseq") %>% 
  with_sample_covariates(c("OS", "sex"))

ggplot(blca.dat, aes(CXCL9, GZMA, color=sex)) +
  geom_point()
```



# Exemplar Analyses

Provide a number of common anlaysis examples using the FacileTCGADataSet.

## Pan-cancer transcriptomics with RPPA

Should we try to reconstruct an analysis from Avi Ma'ayan's recent paper?:
http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005911


# Data Model

The pData data.frame is stored in the sample_covariate table in the internal
database using an [entity-attribute-value model][EAV].

[EAV]: https://en.wikipedia.org/wiki/Entity%E2%80%93attribute%E2%80%93value_model

