---
title: "A unified language for the analysis and visualization of data"
author: "Steve Lianoglou"
date: "16 June 2016"
output: 
  ioslides_presentation:
    transition: faster
    smaller: true
---

```{r setup, include=FALSE}
on.rescomp <- Sys.getenv("HOME") != "/Users/lianogls"
library(rprojroot)
library(JohnnyCache)
library(corrplot)
library(reshape2)
library(magrittr)
library(dplyr)
library(multiGSEA)
library(DT)
library(ggplot2)
library(plotly)
library("FacileRepo")
library(survival)
library(rms)

devtools::load_all("~/workspace/projects/FacileData/apps/TCGA/FacileTCGA")
devtools::load_all("~/workspace/projects/FacileData/apps/atezo/FacileAtezo")
devtools::load_all('~/workspace/projects/FacileData/apps/signatures/CIWG.SigRefine')
if (on.rescomp) {
  library(GenomicsTools)
}
select <- dplyr::select
rename <- dplyr::rename
slice <- dplyr::slice
summarize <- dplyr::summarize

theme_set(theme_bw())

knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE)

slim.df.all <- loadSlimRawMaster() %>%
  group_by(name) %>%
  mutate(N=n()) %>%
  ungroup

sig.df <- bind_rows(
  filter(slim.df.all, sname == 'ER17 ER stress response') %>%
    mutate(name="ER stress response"),
  filter(slim.df.all, name == 'TGFb response (F-TBRS)'),
  filter(slim.df.all, name == 'TGFb response (End-TBRS)'),
  filter(slim.df.all, name == "Cytolytic activity")) %>%
  select(-sname, -signature_type, -protected)

root <- find_root(is_r_package)
r.dir <- file.path(root, 'inst', 'presentations', '2016-06-15-bourgongroup')
c.dir <- init.cache(file.path(r.dir, 'johnnycache'))
source(file.path(r.dir, 'utils.R'))
```

## Problem Statement

<div class="centered">
I routinely want to perform *repeated analyses* over *different subsets of 
samples* across large datasets.
</div>

> 1. I want to **remove the mental burden** of querying for **the same information
     across different datasets** using **dataset specific** identifiers/vocabulary.
> 2. I want **facile access** to subsets of data (genes, samples, samples with
     a mutation in a given pathway, etc.) without having to load an 
     **entire dataset first only to retrieve a small part of it**.
> 3. I want to more easily present the results of these analysis in a **visual and
     interactive way**.

## Examples

> * How do the enrichment scores of my favorite gene signatures correlate with
    each other across a given indication or subtype?
> * Where does a fibroblast-specific TGFb response have any prognostic value?

# Assessment of Signature Activity

## Assessment of Signature Activity

We will assess the correlation structure of the following genes across different
subsets of the TCGA and our atezo clinical trial data:

```{r signature-list, echo=FALSE, results='asis'}
sig.info <- sig.df %>%
  group_by(name) %>%
  summarize(N=n()) %>%
  ungroup
cat(paste(sprintf('*  %s (%d genes)',sig.info$name,sig.info$N), collapse='\n'))
```

## Setup Genesets

We will (of course(!)) us the multiGSEA infrastructure to manage our genesets.

```{r}
(gdb <- GeneSetDb(sig.df))
```

## Score Bladder samples in TCGA (GenomicsTools)

1. Specify samples we want
2. Fetch relevant expression values from samples in (1)
3. Score gene sets on data above

```{r, eval=on.rescomp}
library(GenomicsTools)
blca.all <- loadRNASeqData('BLCA', data='rpkms')                 ## 1 & 2
exprs.gt <- log2(exprs(getTumorData(blca.all)) + 0.1)            ## 1 & 2
rownames(exprs.gt) <- sub('GeneID:', '', rownames(exprs.gt))
scores.gt <- scoreSingleSamples(gdb, exprs.gt, method='svd')     ## 3
```

## Score Bladder samples in TCGA (FacileTCGA)

1. Specify samples we want
2. Fetch relevant expression values from samples in (1)
3. Score gene sets on data above

```{r}
library(FacileTCGA)
tcga <- TcgaDb(cache_size=200000)
samples <- tcga %>%                                              ## 1
  fetch_samples(variable == 'indication' & value == 'BLCA') %>%
  with_sample_covariates('sample_type') %>%
  filter(sample_type == 'tumor')
exprs.ft <- with_expression(samples, sig.df$featureId) %>%       ## 2
  cpm(log=TRUE, prior.count=5)
E <- mcast(exprs.ft, feature_id ~ sample_id, value.var='cpm')
scores.ft <- scoreSingleSamples(gdb, E, method='svd')            ## 3
```

```{r echo=FALSE, eval=!on.rescomp}
scores.gt <- scores.ft
```

## Gene Set Correlation in BLCA {.columns-2}

```{r echo=FALSE, fig.height=3.5, fig.width=4}
plot.cor(scores.gt, title="GenomicsTools")
```

```{r echo=FALSE, fig.height=3.5, fig.width=4.5}
plot.cor(scores.ft, title="FacileTCGA")
```

## What about subtypes?

The datasets in `GenomicsTools` store subtyping information in
indication-specific columns of their respective `pData` objects.

For instance:

* BLCA: no subtyping data stored here, need to fetch from ExpressionPlot
* BRCA
    - PAM50 subtypes in `[['pam50_mrna]]` column (Basal-like, HER2-enriched, 
      Luminal A, Luminal B, Normal-like)
    - Hormone recepter typing: TNBC, ER+/PR+ 
        - Need to parse data in `er_status_by_ihc`, `her2_status_by_ihc`, ...
* COAD/READ: where to we store MSI/MSS info?
* OV: Molcular subtypes in `[[Expression.subtype]]`

<div class="centered">
**High mental burden** to write code that iterates over indications and subtypes
to do something (score gene sets, survival analysis, whatever)
</div>

## Scoring across subtypes within Facile Framework

1. Retrieve the sample grouping information across the TCGA

```{r}
isub.map <- subtype_map(tcga)
```

```{r, echo=FALSE, results='asis'}
## Show only some samples per group: want to keep presentation file size as
## small as possible
isdt <- isub.map %>%
  group_by(indication, subtype) %>%
  mutate(n=n()) %>%
  slice(1:2) %>%
  ungroup %>%
  mutate(indication=factor(indication), subtype=factor(subtype))
datatable(isdt, filter='top')
```

## Scoring across subtypes within Facile Framework

1. Retrieve the sample grouping information across the TCGA
2. Iterate over groups and perform calculation

```{r, calc-scores-foreal, echo=FALSE}
t.start <- Sys.time()
cached(tcga.scores, {
  ## Only running over subset of indications to save time
  ## This 4 seconds over just BLCA subtypes; 1.5 mins over BLCA, OV, and CRC
  isub.map %>%
    filter(indication %in% c('BLCA', 'OV', 'CRC')) %>%
    group_by(indication, subtype, sample_type) %>%
    do({
      with_expression(., sig.df$featureId, db=tcga) %>%
        cpm(log=TRUE, prior.count=5) %>%
        mcast(feature_id ~ sample_id, value.var='cpm') %>%
        scoreSingleSamples(gdb, ., method='svd')
    }) %>%
    ungroup
}, 'facile-tcga-gs-scores.rds', force=FALSE)
t.end <- Sys.time()
```

```{r, calc-scores-smoke-mirrors, eval=FALSE}
## Runs in 4 seconds over just bladder subtypes, 1.5 mins for BLCA, OV, CRC
tcga.scores <- isub.map %>%
  filter(indication %in% c('BLCA', 'OV', 'CRC')) %>%
  group_by(indication, subtype, sample_type) %>%
  do({
    with_expression(., sig.df$featureId, db=tcga) %>%
      cpm(log=TRUE, prior.count=5) %>%
      mcast(feature_id ~ sample_id, value.var='cpm') %>%
      scoreSingleSamples(gdb, ., method='svd')
  }) %>%
  ungroup
```

## Signature correlations across subtypes {.columns-2}

```{r echo=FALSE, fig.height=3.5, fig.width=4}
plot.cor(tcga.scores, 'BLCA', 'all', title="BLCA (all)", order="original")
```

```{r echo=FALSE, fig.height=3.5, fig.width=4.5}
plot.cor(tcga.scores, 'BLCA', 'basal', title="BLCA (basal)", order="original")
```

## Visualizing scores across analysis

```{r}
tcga.scores <- tcga.scores %>%
  group_by(name) %>%
  mutate(score.center=score - median(score),
         isub=paste(indication, subtype, sep='_')) %>%
  ungroup

gg <- ggplot(filter(tcga.scores, sample_type == 'tumor'),
             aes(isub, score.center, fill=name)) +
  geom_boxplot() +
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5))
gg
```

## Visualizing scores across analysis (interactively)

`ggplotly` isn't *quite* there yet

```{r}
ggplotly(gg)
```

## Are these signatures prognostic in our data?

We could have used the same code structure to run a survival analysis "on the
fly", but since we already have our sample/subtype specific scores in hand,
let's decorate them with survival information and run a survival analysis per
subtype.

```{r, echo=FALSE}
## we dropped dataset by accident and casting for geneset scoring changed
## sample_id to sample
xref <- match(tcga.scores$sample, isub.map$sample_id)
tcga.scores %<>% mutate(dataset=isub.map$dataset[xref], sample_id=sample)
```

```{r}
with.surv <- tcga.scores %>%
  with_sample_covariates(c('OS', 'stage'), db=tcga)

survs <- with.surv %>%
  filter(sample_type == 'tumor') %>%
  group_by(indication, subtype, name) %>%
  do(cox={
    sig.status <- ifelse(.$score > median(.$score), 'high', 'low')
    sig.status <- factor(sig.status, c('low', 'high'))
    srv <- Surv(.$tte_OS, event=.$event_OS)
    cph(srv ~ strat(sig.status), surv=TRUE, x=TRUE, y=TRUE)
  }) %>%
  ungroup

cox <- filter(survs, indication=='BLCA', subtype == 'all',
              name == 'TGFb response (F-TBRS)') %>%
  extract2('cox')
cox <- cox[[1]]

survplot(cox, sig.status=c('low', 'high'),
         grid=NULL, conf.int=FALSE,
         n.risk=TRUE, y.n.risk='auto', xlab=NULL)
```

# Atezo

## Scoring code is the same

We can do same analysis, same code, different back end

```{r, eval=FALSE}
atezo <- AtezoDb()
## asub.map <- subtype_map(atezo)
asub.map <- atezo_isub_map(atezo)
atezo.scores <- isub.map %>%
  filter(indication %in% 'bladder') %>%
  group_by(indication, subtype, sample_type) %>%
  do({
    with_expression(., sig.df$featureId, db=tcga) %>%
      cpm(log=TRUE, prior.count=5) %>%
      mcast(feature_id ~ sample_id, value.var='cpm') %>%
      scoreSingleSamples(gdb, ., method='svd')
  }) %>%
  ungroup
```

## So ... ?

Data are imported into a database to allow for "facile" access
of genes and samples based on gene identity and sample characteristics. Data
are spread across a number of "long" tables:

* `expression`: stores gene expression counts (`melt(exprs(es))`)
* `gene_info`: stores metadata about genes (`fData(es)`)
* `sample_covariate`: stores sample covariate information in a "knowledgable"
  way.
* `sample_stats`: lib.size, norm.factors, things like that.

These tables are queried and joined to get to the specific chunk of data you
want.

## Tables

```{r}
head(gene_info_tbl(tcga), n=3)
head(expression_tbl(tcga), n=3)
head(sample_stats_tbl(tcga), n=3)
```

## Covariates endowed with knowledge

Covariates name, value, class, and type are stored into the database:

```{r}
head(sample_covariate_tbl(tcga))
```

These are all of the things the database knows about the TCGA, and now you know
that the database knows ... and the database knows that you know that it knows ...

```{r echo=FALSE, results='asis'}
cov.info <- sample_covariate_tbl(tcga) %>%
  collect(n=Inf) %>%
  group_by(variable) %>%
  summarize(n_categories=if (type[1L] == 'categorical') {
    length(unique(value))
  } else Inf) %>%
  ungroup
knitr::kable(cov.info)
```








# Evolving Language

## Score Bladder samples in TCGA (FacileTCGA)

1. Select samples we want
2. Fetch relevant expression values from samples in (1)
3. Score gene sets on data above

```{r, eval=FALSE}
scores.ft <- tcga %>%                                              ## 1
  fetch_samples(variable == 'indication' & value == 'BLCA') %>%
  with_sample_covariates('sample_type') %>%
  filter(sample_type == 'tumor') %>%
  with_expression(sig.df$featureId) %>%                   ## 2
  cpm(log=TRUE, prior.count=5) %>%
  mcast(feature_id ~ sample_id, value.var='cpm') %>%
  scoreSingleSamples(gdb, ., method='svd')                         ## 3
```

## ... eventually

```{r, eval=FALSE}
scores.ft <- tcga %>%                                              ## 1
  filter_samples(indication == 'BLCA' & sample_type == 'tumor') %>%
  with_expression(sig.df$featureId) %>%                            ## 2
  cpm(log=TRUE, prior.count=5) %>%
  mcast(feature_id ~ sample_id, value.var='cpm') %>%
  scoreSingleSamples(gdb, ., method='svd')                         ## 3
```


  


